torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -463,6 +470,10 @@ void Reducer::finalize_backward() {
   AT_ASSERT(expect_autograd_hooks_);
```

Comment:
`expect_autograd_hooks_` was set to true in `prepare_for_forward`.  Should it actually be false if all params are unused?

Reply:
In `prepare_for_backward`. I think it is fine as is, because `finalize_backward` will be called inline if all parameters are unused. Essentially a set to true followed by a set to false. Adding more conditionals (and e.g. a `all_parameters_unused_` flag) would make things more fragile IMO.